packages/trimerge-sync/src/TrimergeClient.ts

```diff
@@ -310,20 +310,39 @@ export class TrimergeClient<
       ),
   };
 
-  getCommitDoc(ref: string): CommitDoc<SavedDoc, CommitMetadata> {
-    const doc = this.docCache.get(ref);
-    if (doc !== undefined) {
-      return doc;
+  getCommitDoc(headRef: string): CommitDoc<SavedDoc, CommitMetadata> {
+    let baseDoc: CommitDoc<SavedDoc, CommitMetadata> | undefined;
+    const commitWalk: string[] = [];
+    let currentCommitRef: string | undefined = headRef;
+    while (!baseDoc && currentCommitRef !== undefined) {
+      if (this.docCache.has(currentCommitRef)) {
+        baseDoc = this.docCache.get(currentCommitRef);
+        break;
+      }
+
+      commitWalk.push(currentCommitRef);
+      const commit = this.getCommit(currentCommitRef);
+      currentCommitRef = commit.baseRef;
+    }
+
+    let resultDoc = baseDoc;
+    for (let i = commitWalk.length - 1; i >= 0; i--) {
+      const { ref, delta, metadata } = this.getCommit(commitWalk[i]);
```

Comment:
Yours is probably faster but this seems easier to reason about imo:
```suggestion
    while (commitWalk.length > 0) {
      const { ref, delta, metadata } = this.getCommit(commitWalk.pop());
```

Reply:
Ack, but will leave as-is since it seems close in terms of understandability.